full-node/sov-stf-runner/src/prover_service/parallel/prover.rs

```diff
@@ -0,0 +1,178 @@
+}
+
+impl<StateRoot, Witness, Da: DaSpec> ProverState<StateRoot, Witness, Da> {
+    fn remove(&mut self, hash: &Hash) -> Option<ProverStatus<StateRoot, Witness, Da>> {
+        self.prover_status.remove(hash)
+    }
+
+    fn set_to_proving(&mut self, hash: Hash) -> Option<ProverStatus<StateRoot, Witness, Da>> {
+        self.prover_status.insert(hash, ProverStatus::Proving)
+    }
+
+    fn set_to_proved(
+        &mut self,
+        hash: Hash,
+        proof: Result<Proof, anyhow::Error>,
+    ) -> Option<ProverStatus<StateRoot, Witness, Da>> {
+        match proof {
+            Ok(p) => self.prover_status.insert(hash, ProverStatus::Proved(p)),
+            Err(e) => self.prover_status.insert(hash, ProverStatus::Err(e)),
+        }
+    }
+
+    fn get_prover_status(&self, hash: Hash) -> Option<&ProverStatus<StateRoot, Witness, Da>> {
+        self.prover_status.get(&hash)
+    }
+}
+
+pub(crate) struct Prover<StateRoot, Witness, Da: DaService> {
+    prover_state: Arc<Mutex<ProverState<StateRoot, Witness, Da::Spec>>>,
+}
+
+impl<StateRoot, Witness, Da> Prover<StateRoot, Witness, Da>
+where
+    Da: DaService,
+    StateRoot: Serialize + DeserializeOwned + Clone + AsRef<[u8]> + Send + Sync + 'static,
+    Witness: Serialize + DeserializeOwned + Send + Sync + 'static,
+{
+    pub(crate) fn new() -> Self {
+        Self {
+            prover_state: Arc::new(Mutex::new(ProverState {
+                prover_status: Default::default(),
+            })),
+        }
+    }
+
+    pub(crate) fn submit_witness(
+        &self,
+        state_transition_data: StateTransitionData<StateRoot, Witness, Da::Spec>,
+    ) {
+        let header_hash = state_transition_data.da_block_header.hash().into();
+        let data = ProverStatus::WitnessSubmitted(state_transition_data);
+
+        self.prover_state
+            .lock()
+            .expect("Lock was poisoned")
+            .prover_status
+            .insert(header_hash, data);
+    }
+
+    pub(crate) fn start_proving<Vm, V>(
+        &self,
+        block_header_hash: Hash,
+        config: Arc<ProofGenConfig<V, Da, Vm>>,
+        mut vm: Vm,
+        zk_storage: V::PreState,
+    ) -> Result<(), ProverServiceError>
+    where
+        Vm: ZkvmHost + 'static,
+        V: StateTransitionFunction<Vm::Guest, Da::Spec> + Send + Sync + 'static,
+        V::PreState: Send + Sync + 'static,
+    {
+        let prover_state_clone = self.prover_state.clone();
+
+        let mut prover_state = self.prover_state.lock().expect("Lock was poisoned");
+
+        let prover_status = prover_state
+            .remove(&block_header_hash)
+            .ok_or_else(|| anyhow::anyhow!("Missing status for {:?}", block_header_hash))?;
+
+        match prover_status {
+            ProverStatus::WitnessSubmitted(state_tranistion_data) => {
+                prover_state.set_to_proving(block_header_hash);
+                vm.add_hint(state_tranistion_data);
+
+                rayon::spawn(move || {
```

Comment:
Yes, this is point `1` in the PR description. I left it as a follow-up because the code became a bit complicated, and currently, there is no good way to test it, so I don't want to add even more complexity before #1165  is fixed. 
The following PR will be #1165, and after that, I will add handling for the busy status.

Reply:
Sounds good, we can also take this topic on the other item

https://github.com/Sovereign-Labs/sovereign-sdk/pull/1160/files#r1399167455